nx/source/runtime/devices/usb_comms.c

```diff
@@ -25,69 +26,127 @@ static usbCommsInterface g_usbCommsInterfaces[TOTAL_INTERFACES];
 
 static RwLock g_usbCommsLock;
 
-static Result _usbCommsInterfaceInit(usbCommsInterface *interface, u8 bInterfaceClass, u8 bInterfaceSubClass, u8 bInterfaceProtocol);
+static Result _usbCommsInterfaceInit1x(u32 intf_ind);
+static Result _usbCommsInterfaceInit5x(u32 intf_ind);
+static Result _usbCommsInterfaceInit(u32 intf_ind);
 
 static Result _usbCommsWrite(usbCommsInterface *interface, const void* buffer, size_t size, size_t *transferredSize);
 
-Result usbCommsInitializeEx(u32 *interface, u8 bInterfaceClass, u8 bInterfaceSubClass, u8 bInterfaceProtocol)
+Result usbCommsInitializeEx(u32 num_interfaces)
 {
-    bool found=0;
-    usbCommsInterface *inter = NULL;
-
+    Result rc = 0;
     rwlockWriteLock(&g_usbCommsLock);
-
-    if (g_usbCommsInitialized && interface==NULL) {
-        rwlockWriteUnlock(&g_usbCommsLock);
-        return 0;
-    }
-
-    Result rc=0;
-    u32 i = 0;
-
-    if (!g_usbCommsInitialized) rc = usbDsInitialize(UsbComplexId_Default, NULL);
-
-    if (R_SUCCEEDED(rc)) {
-        for(i=0; i<TOTAL_INTERFACES; i++) {
-            inter = &g_usbCommsInterfaces[i];
-            rwlockReadLock(&inter->lock);
-            if (!inter->initialized) found=1;
-            rwlockReadUnlock(&inter->lock);
-
-            if (found) break;
+    
+    if (g_usbCommsInitialized) {
+        rc = MAKERESULT(Module_Libnx, LibnxError_AlreadyInitialized); 
+    } else if (num_interfaces > TOTAL_INTERFACES) {
+        rc = MAKERESULT(Module_Libnx, LibnxError_OutOfMemory);
+    } else {
+        rc = usbDsInitialize();
+        
+        if (R_SUCCEEDED(rc)) {
+            
+            if (kernelAbove500()) {
+                u8 iManufacturer, iProduct, iSerialNumber;
+                static const u16 supported_langs[1] = {0x0409};
+                // Send language descriptor
+                rc = usbDsAddUsbLanguageStringDescriptor(NULL, supported_langs, sizeof(supported_langs)/sizeof(u16));
+                // Send manufacturer
+                if (R_SUCCEEDED(rc)) rc = usbDsAddUsbStringDescriptor(&iManufacturer, "Switchbrew");
+                // Send product
+                if (R_SUCCEEDED(rc)) rc = usbDsAddUsbStringDescriptor(&iProduct, "libnx USB comms");
```

Comment:
Should a different product string be sent here?

Reply:
Change it to what <5.0 uses.